### PR TITLE
Address Stint shortcomings

### PR DESCRIPTION
**Pull request by mratsim**
This should hopefully address Stint shortcomings mentioned in our discussion.

- No template magic to create the public API procs. This also allow us to add documentation
- Tests that were not enabled (modular arithmetics in particular) are now enabled
- Adding a `toUint` proc, fixing #58, that returns the raw least significant word
- Adding `toUint64` and `toInt64`, fixing #48, that return the either the raw least significant word on 64-bit machine or least significant 2 words on a 32-bit machine.
- Following upstream fixes, solve #40 and us static default argument for `parse` (base 10), `toString` (base 10) and `dumpHex` (bigEndian) instead of overloaded proc
- deprecate `significantBytesBE` as it works on `openarray[bytes]` and should instead be in https://github.com/status-im/nim-byteutils. Note that for blobs coming from stint, assuming Uint256 `255 - x.countLeadingZeroBits` should be used instead

### CODE REVIEW

**Comment by mratsim** [comment]
## Showstopper Issues:

```
$ nim -v
Nim Compiler Version 0.19.0 [MacOSX: amd64]
Compiled at 2018-09-26
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: f6c5c636bb1a1f4e1301ae0ba5a8afecef439132
active boot switches: -d:release
```

See for deactivated tests: https://github.com/status-im/nim-stint/pull/66/files#diff-1fb444f4150eadb60862eb713582b5fe

I currently have to comment out the following tests:

- test_uint_comparison:
https://github.com/status-im/nim-stint/blob/c6fc6b8a304cdef54bc739000cd8b5678d956ad3/tests/test_uint_comparison.nim#L30-L37
```Nim
test_uint_comparison.nim(12, 7) template/generic instantiation from here
test_uint_comparison.nim(30, 8) template/generic instantiation from here
test_uint_comparison.nim(31, 5) template/generic instantiation from here
test_uint_comparison.nim(34, 17) template/generic instantiation from here
../../../../.choosenim/toolchains/nim-0.19.0/lib/pure/unittest.nim(654, 15) template/generic instantiation from here
../../../../.choosenim/toolchains/nim-0.19.0/lib/pure/unittest.nim(551, 20) Error: generic instantiation too nested
```

- test_io and test_uint_muldiv:
Line 22
https://github.com/status-im/nim-stint/blob/c6fc6b8a304cdef54bc739000cd8b5678d956ad3/stint/private/uint_div.nim#L206-L227
```Nim
../stint/private/uint_div.nim(222, 10) Error: type mismatch: got <UintImpl[system.uint32], UintImpl[system.uint32]>
but expected one of:
proc `<=`[Enum: enum](x, y: Enum): bool
  first type mismatch at position: 1
  required type: Enum: enum
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: string): bool
  first type mismatch at position: 1
  required type: string
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: char): bool
  first type mismatch at position: 1
  required type: char
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[T](x, y: set[T]): bool
  first type mismatch at position: 1
  required type: set[T]
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: bool): bool
  first type mismatch at position: 1
  required type: bool
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[T](x, y: ref T): bool
  first type mismatch at position: 1
  required type: ref T
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: pointer): bool
  first type mismatch at position: 1
  required type: pointer
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: float32): bool
  first type mismatch at position: 1
  required type: float32
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: int): bool
  first type mismatch at position: 1
  required type: int
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: float): bool
  first type mismatch at position: 1
  required type: float
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[T: tuple](x, y: T): bool
  first type mismatch at position: 1
  required type: T: tuple
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: int8): bool
  first type mismatch at position: 1
  required type: int8
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: int32): bool
  first type mismatch at position: 1
  required type: int32
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[T: SomeUnsignedInt](x, y: T): bool
  first type mismatch at position: 1
  required type: T: SomeUnsignedInt
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: int16): bool
  first type mismatch at position: 1
  required type: int16
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`(x, y: int64): bool
  first type mismatch at position: 1
  required type: int64
  but expression 'd' is of type: UintImpl[system.uint32]
func `<=`(x, y: IntImpl): bool
  first type mismatch at position: 1
  required type: IntImpl
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[bits: static[int]](a167546: StUint[bits]; b167548: int{lit}): bool
  first type mismatch at position: 1
  required type: StUint[<=.bits]
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[bits: static[int]](a167550: int{lit}; b167552: StUint[bits]): bool
  first type mismatch at position: 1
  required type: int
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[bits: static[int]](a167558: int{lit}; b167560: StInt[bits]): bool
  first type mismatch at position: 1
  required type: int
  but expression 'd' is of type: UintImpl[system.uint32]
proc `<=`[bits: static[int]](a167554: StInt[bits]; b167556: int{lit}): bool
  first type mismatch at position: 1
  required type: StInt[<=.bits]
  but expression 'd' is of type: UintImpl[system.uint32]
func `<=`(x161156, y161158: StInt): bool
  first type mismatch at position: 1
  required type: StInt
  but expression 'd' is of type: UintImpl[system.uint32]
func `<=`(x, y: StUint): bool
  first type mismatch at position: 1
  required type: StUint
  but expression 'd' is of type: UintImpl[system.uint32]

expression: d <= r
```

**Comment by tersec** [comment]
Fixes for https://github.com/status-im/nim-stint/issues/58 and https://github.com/status-im/nim-stint/issues/48 are quite nice to see, and address the most acute concrete issues that have come up in VM implementation. Section 9.2 of the yellow paper specifically mentions:
>  Due to this fee it is highly unlikely addresses will ever go above 32-bit  bounds. That said, implementations must be able to manage this eventuality.
So on 32-bit platforms, `toUint64` is also useful.

File: stint/private/datatypes.nim [hunk 1]
```diff
@@ -10,89 +10,147 @@
...
+#   - Note that while shift-right and left can easily be done an array of bytes
+#     this would have reduced performance compared to moving 64-bit words.
+#     An efficient implementation on array of words would require checking the shift
+#     versus a half-word to deal with carry-in/out from and to the adjacent words
+#     similar to a recursive implementation.
+#
+# Iterations over the whole integers, for example for `==` is always unrolled.
+# Due to being on the stack, any optimizing compiler should compile that to efficient memcmp
+
+when not defined(stint_test):
```
Comment (mratsim) [inline stint/private/datatypes.nim]: Here I just exchanged the release and test backend so that the release structure is presented first

**Comment by coffeepots** [comment]
Some really nice changes in here. Particularly dropping the template magic makes the code much easier to read IMO, and `toUInt` & `toUInt64` will be extremely useful.

**Comment by mratsim** [comment]
This should address all leftover comments. Please note that the change from default name `base` to `radix` is breaking (but easy to fix).

I didn't see any usage of `parse` or `toString` with a named `base` parameter in Nimbus though so it should be painless.

File: stint/io.nim [hunk 2]
```diff
@@ -73,9 +73,22 @@ func to*(x: SomeInteger, T: typedesc[Stint]): T =
 func to*(x: SomeUnsignedInt, T: typedesc[StUint]): T =
   stuint(x, result.bits)
 
-func toInt*(num: Stint or StUint): int {.inline.}=
-  # Returns as int. Result is modulo 2^(sizeof(int)
-  num.data.least_significant_word.int
+func truncate*(num: Stint or StUint, T: typedesc[int or uint or int64 or uint64]): T {.inline.}=
+  ## Extract the int, uint, int64 or uint64 portion of a multi-precision integer.
+  ## Note that int and uint are 32-bit on 32-bit platform.
```
Comment (arnetheduck) [inline stint/io.nim]: just `platform-dependent` - could  be 16 or 128!

**Pull request merged**